File: src/hypeedge/backtest/walk_forward.py
"""Walk-forward analysis and anti-overfitting tools (design doc §6.1).

Implements:
- Walk-forward analysis with rolling train/validation windows
- Monte Carlo bootstrap simulation for statistical significance
- Bonferroni multiple testing correction
"""

from __future__ import annotations

import math
import random
from dataclasses import dataclass
from typing import Any

import structlog

from hypeedge.backtest.broker import FeeConfig, SlippageConfig, SlippageMode
from hypeedge.backtest.engine import BacktestEngine, BacktestResult, StrategyFactory
from hypeedge.core.models import Candle, FundingRate
from hypeedge.core.types import Usd

logger = structlog.get_logger(__name__)

_HOUR_MS = 3_600_000
_DAY_MS = 24 * _HOUR_MS


@dataclass(frozen=True)
class WalkForwardWindow:
    """Result of a single walk-forward window."""

    window_index: int
    train_start_ms: int
    train_end_ms: int
    validate_start_ms: int
    validate_end_ms: int
    train_result: BacktestResult
    validate_result: BacktestResult


@dataclass(frozen=True)
class WalkForwardResult:
    """Aggregate result of walk-forward analysis."""

    windows: list[WalkForwardWindow]
    aggregate_return_pct: float
    aggregate_sharpe: float
    aggregate_max_drawdown_pct: float
    total_validate_trades: int
    n_windows: int

    def to_dict(self) -> dict[str, Any]:
        return {
            "n_windows": self.n_windows,
            "aggregate_return_pct": round(self.aggregate_return_pct, 4),
            "aggregate_sharpe": round(self.aggregate_sharpe, 4),
            "aggregate_max_drawdown_pct": round(self.aggregate_max_drawdown_pct, 4),
            "total_validate_trades": self.total_validate_trades,
            "windows": [
                {
                    "index": w.window_index,
                    "train_return_pct": round(w.train_result.metrics.total_return_pct, 4),
                    "validate_return_pct": round(w.validate_result.metrics.total_return_pct, 4),
                    "validate_sharpe": round(w.validate_result.metrics.sharpe_ratio, 4),
                    "validate_max_dd": round(w.validate_result.metrics.max_drawdown_pct, 4),
                }
                for w in self.windows
            ],
        }


@dataclass(frozen=True)
class MonteCarloResult:
    """Result of Monte Carlo bootstrap simulation."""

    n_simulations: int
    return_ci_lower: float
    return_ci_upper: float
    sharpe_ci_lower: float
    sharpe_ci_upper: float
    drawdown_ci_lower: float
    drawdown_ci_upper: float
    p_value_return: float  # Probability of achieving observed return by chance
    observed_return: float
    observed_sharpe: float

    def to_dict(self) -> dict[str, Any]:
        return {
            "n_simulations": self.n_simulations,
            "return_ci": [round(self.return_ci_lower, 4), round(self.return_ci_upper, 4)],
            "sharpe_ci": [round(self.sharpe_ci_lower, 4), round(self.sharpe_ci_upper, 4)],
            "drawdown_ci": [round(self.drawdown_ci_lower, 4), round(self.drawdown_ci_upper, 4)],
            "p_value_return": round(self.p_value_return, 4),
            "observed_return": round(self.observed_return, 4),
            "observed_sharpe": round(self.observed_sharpe, 4),
        }


class WalkForwardEngine:
    """Walk-forward analysis engine (design doc §6.1).

    Splits data into rolling train/validation windows. Parameters are
    optimized on training segments and evaluated on validation segments.
    This better simulates real parameter decay than a single train/test split.
    """

    def __init__(
        self,
        engine: BacktestEngine | None = None,
        fee_config: FeeConfig | None = None,
        slippage_config: SlippageConfig | None = None,
    ) -> None:
        self._engine = engine or BacktestEngine(fee_config, slippage_config)

    async def run_walk_forward(
        self,
        candles: list[Candle],
        funding_rates: list[FundingRate] | None,
        strategy_factory: StrategyFactory,
        train_days: int = 60,
        validate_days: int = 30,
        step_days: int = 30,
        initial_capital: Usd | None = None,
        slippage_mode: SlippageMode = SlippageMode.PESSIMISTIC,
    ) -> WalkForwardResult:
        """Run walk-forward analysis with rolling windows.

        Args:
            candles: Historical candle data sorted by timestamp.
            funding_rates: Historical funding data (optional).
            strategy_factory: Creates strategy instances for each window run.
            train_days: Training window length in days.
            validate_days: Validation window length in days.
            step_days: Window slide step in days.
            initial_capital: Starting capital per window.
            slippage_mode: Fill simulation mode.

        Returns:
            WalkForwardResult with per-window and aggregate metrics.
        """
        if initial_capital is None:
            initial_capital = Usd(10_000.0)

        if not candles:
            logger.warning("walk_forward_empty_candles")
            return WalkForwardResult(
                windows=[],
                aggregate_return_pct=0.0,
                aggregate_sharpe=0.0,
                aggregate_max_drawdown_pct=0.0,
                total_validate_trades=0,
                n_windows=0,
            )

        train_ms = train_days * _DAY_MS
        validate_ms = validate_days * _DAY_MS
        step_ms = step_days * _DAY_MS

        first_ts: int = int(candles[0].timestamp)
        last_ts: int = int(candles[-1].timestamp)
        total_span = last_ts - first_ts

        logger.info(
            "walk_forward_start",
            train_days=train_days,
            validate_days=validate_days,
            step_days=step_days,
            total_candles=len(candles),
            total_span_days=total_span / _DAY_MS,
        )

        windows: list[WalkForwardWindow] = []
        window_start = first_ts
        window_idx = 0

        while True:
            train_start = window_start
            train_end = train_start + train_ms
            validate_start = train_end
            validate_end = validate_start + validate_ms

            # Check if we have enough data for this window
            if validate_end > last_ts:
                break

            # Slice candles for train and validate
            train_candles = [c for c in candles if train_start <= c.timestamp < train_end]
            validate_candles = [c for c in candles if validate_start <= c.timestamp < validate_end]

            if not train_candles or not validate_candles:
                window_start += step_ms
                window_idx += 1
                continue

            # Slice funding rates
            train_funding = self._slice_funding(funding_rates, train_start, train_end)
            validate_funding = self._slice_funding(funding_rates, validate_start, validate_end)

            logger.debug(
                "walk_forward_window",
                window=window_idx,
                train_candles=len(train_candles),
                validate_candles=len(validate_candles),
            )

            # Run on training segment
            train_result = await self._engine.run(
                candles=train_candles,
                funding_rates=train_funding,
                strategy_factory=strategy_factory,
                initial_capital=initial_capital,
                slippage_mode=slippage_mode,
            )

            # Run on validation segment
            validate_result = await self._engine.run(
                candles=validate_candles,
                funding_rates=validate_funding,
                strategy_factory=strategy_factory,
                initial_capital=initial_capital,
                slippage_mode=slippage_mode,
            )

            windows.append(
                WalkForwardWindow(
                    window_index=window_idx,
                    train_start_ms=train_start,
                    train_end_ms=train_end,
                    validate_start_ms=validate_start,
                    validate_end_ms=validate_end,
                    train_result=train_result,
                    validate_result=validate_result,
                )
            )

            window_start += step_ms
            window_idx += 1

        # Compute aggregate metrics
        if windows:
            avg_return = sum(w.validate_result.metrics.total_return_pct for w in windows) / len(windows)
            avg_sharpe = sum(w.validate_result.metrics.sharpe_ratio for w in windows) / len(windows)
            max_dd = max(w.validate_result.metrics.max_drawdown_pct for w in windows)
            total_trades = sum(w.validate_result.metrics.trade_count for w in windows)
        else:
            avg_return = 0.0
            avg_sharpe = 0.0
            max_dd = 0.0
            total_trades = 0

        result = WalkForwardResult(
            windows=windows,
            aggregate_return_pct=avg_return,
            aggregate_sharpe=avg_sharpe,
            aggregate_max_drawdown_pct=max_dd,
            total_validate_trades=total_trades,
            n_windows=len(windows),
        )

        logger.info(
            "walk_forward_complete",
            n_windows=result.n_windows,
            aggregate_return=f"{avg_return:.4%}",
            aggregate_sharpe=f"{avg_sharpe:.2f}",
        )
        return result

    @staticmethod
    def _slice_funding(
        rates: list[FundingRate] | None,
        start_ms: int,
        end_ms: int,
    ) -> list[FundingRate] | None:
        if not rates:
            return None
        return [r for r in rates if start_ms <= r.timestamp < end_ms]


def run_monte_carlo(
    equity_curve: list[tuple[int, float]],
    n_simulations: int = 1000,
    confidence: float = 0.95,
    seed: int | None = None,
) -> MonteCarloResult:
    """Monte Carlo bootstrap simulation (design doc §6.1).

    Randomly reshuffles (bootstrap) historical return series to test
    whether strategy returns are significantly better than random.

    Args:
        equity_curve: List of (timestamp, equity) tuples.
        n_simulations: Number of bootstrap iterations.
        confidence: Confidence level for intervals (e.g. 0.95 for 95%).
        seed: Random seed for reproducibility.

    Returns:
        MonteCarloResult with confidence intervals and p-value.
    """
    if len(equity_curve) < 2:
        return MonteCarloResult(
            n_simulations=0,
            return_ci_lower=0.0,
            return_ci_upper=0.0,
            sharpe_ci_lower=0.0,
            sharpe_ci_upper=0.0,
            drawdown_ci_lower=0.0,
            drawdown_ci_upper=0.0,
            p_value_return=1.0,
            observed_return=0.0,
            observed_sharpe=0.0,
        )

    rng = random.Random(seed)

    # Compute observed returns
    observed_returns = _compute_returns(equity_curve)
    observed_total_return = (equity_curve[-1][1] / equity_curve[0][1]) - 1.0 if equity_curve[0][1] > 0 else 0.0
    observed_sharpe = _compute_sharpe(observed_returns)

    # Bootstrap simulations
    sim_returns: list[float] = []
    sim_sharpes: list[float] = []
    sim_drawdowns: list[float] = []

    for _ in range(n_simulations):
        # Resample returns with replacement
        resampled = [rng.choice(observed_returns) for _ in range(len(observed_returns))]
        # Reconstruct equity from resampled returns
        sim_equity = equity_curve[0][1]
        sim_equity_curve = [(equity_curve[0][0], sim_equity)]
        for i, ret in enumerate(resampled):
            sim_equity = sim_equity * (1.0 + ret)
            ts = equity_curve[min(i + 1, len(equity_curve) - 1)][0]
            sim_equity_curve.append((ts, sim_equity))

        initial = sim_equity_curve[0][1]
        sim_total_return = (sim_equity_curve[-1][1] / initial) - 1.0 if initial > 0 else 0.0
        sim_returns.append(sim_total_return)
        sim_sharpes.append(_compute_sharpe(resampled))
        sim_drawdowns.append(_compute_max_drawdown(sim_equity_curve))

    # Compute confidence intervals
    alpha = (1.0 - confidence) / 2.0
    sim_returns.sort()
    sim_sharpes.sort()
    sim_drawdowns.sort()

    lower_idx = max(0, int(alpha * n_simulations))
    upper_idx = min(n_simulations - 1, int((1.0 - alpha) * n_simulations))

    # P-value: fraction of simulations that beat the observed return
    p_value = sum(1 for r in sim_returns if r >= observed_total_return) / n_simulations

    return MonteCarloResult(
        n_simulations=n_simulations,
        return_ci_lower=sim_returns[lower_idx],
        return_ci_upper=sim_returns[upper_idx],
        sharpe_ci_lower=sim_sharpes[lower_idx],
        sharpe_ci_upper=sim_sharpes[upper_idx],
        drawdown_ci_lower=sim_drawdowns[lower_idx],
        drawdown_ci_upper=sim_drawdowns[upper_idx],
        p_value_return=p_value,
        observed_return=observed_total_return,
        observed_sharpe=observed_sharpe,
    )


def bonferroni_correction(p_value: float, n_tests: int) -> float:
    """Apply Bonferroni multiple testing correction.

    Design doc §6.1: "100 个参数组合测试意味着显著性阈值从 0.05 上升到 0.0005"
    """
    return min(p_value * n_tests, 1.0)


def _compute_returns(equity_curve: list[tuple[int, float]]) -> list[float]:
    """Compute simple returns from an equity curve."""
    returns: list[float] = []
    for i in range(1, len(equity_curve)):
        prev = equity_curve[i - 1][1]
        curr = equity_curve[i][1]
        if prev > 0:
            returns.append((curr - prev) / prev)
    return returns


def _compute_sharpe(returns: list[float]) -> float:
    """Compute annualized Sharpe from a return series."""
    if not returns:
        return 0.0
    mean_ret = sum(returns) / len(returns)
    variance = sum((r - mean_ret) ** 2 for r in returns) / len(returns)
    std_ret = math.sqrt(variance) if variance > 0 else 0.0
    if std_ret == 0:
        return 0.0
    # Assume hourly returns → ~8760 per year
    return (mean_ret / std_ret) * math.sqrt(8760)


def _compute_max_drawdown(equity_curve: list[tuple[int, float]]) -> float:
    """Compute max drawdown from an equity curve."""
    peak = 0.0
    max_dd = 0.0
    for _, eq in equity_curve:
        if eq > peak:
            peak = eq
        if peak > 0:
            dd = (peak - eq) / peak
            if dd > max_dd:
                max_dd = dd
    return max_dd

File: crates/trading/src/backtest/walk_forward.rs
//! Walk-forward analysis and anti-overfitting tools, port of
//! `src/hypeedge/backtest/walk_forward.py`.

/// Result of a single walk-forward window.
#[derive(Debug, Clone)]
pub struct WalkForwardWindow {
    pub window_index: usize,
    pub train_start_ms: i64,
    pub train_end_ms: i64,
    pub validate_start_ms: i64,
    pub validate_end_ms: i64,
    pub train_total_return_pct: f64,
    pub validate_total_return_pct: f64,
    pub validate_sharpe: f64,
    pub validate_max_drawdown_pct: f64,
    pub validate_trade_count: usize,
}

/// Aggregate result of walk-forward analysis.
#[derive(Debug, Clone)]
pub struct WalkForwardResult {
    pub windows: Vec<WalkForwardWindow>,
    pub aggregate_return_pct: f64,
    pub aggregate_sharpe: f64,
    pub aggregate_max_drawdown_pct: f64,
    pub total_validate_trades: usize,
    pub n_windows: usize,
}

impl WalkForwardResult {
    pub fn to_dict(&self) -> serde_json::Value {
        serde_json::json!({
            "n_windows": self.n_windows,
            "aggregate_return_pct": round4(self.aggregate_return_pct),
            "aggregate_sharpe": round4(self.aggregate_sharpe),
            "aggregate_max_drawdown_pct": round4(self.aggregate_max_drawdown_pct),
            "total_validate_trades": self.total_validate_trades,
            "windows": self.windows.iter().map(|w| serde_json::json!({
                "index": w.window_index,
                "train_return_pct": round4(w.train_total_return_pct),
                "validate_return_pct": round4(w.validate_total_return_pct),
                "validate_sharpe": round4(w.validate_sharpe),
                "validate_max_dd": round4(w.validate_max_drawdown_pct),
            })).collect::<Vec<_>>(),
        })
    }
}

/// Result of Monte Carlo bootstrap simulation.
#[derive(Debug, Clone)]
pub struct MonteCarloResult {
    pub n_simulations: usize,
    pub return_ci_lower: f64,
    pub return_ci_upper: f64,
    pub sharpe_ci_lower: f64,
    pub sharpe_ci_upper: f64,
    pub drawdown_ci_lower: f64,
    pub drawdown_ci_upper: f64,
    pub p_value_return: f64,
    pub observed_return: f64,
    pub observed_sharpe: f64,
}

impl MonteCarloResult {
    pub fn to_dict(&self) -> serde_json::Value {
        serde_json::json!({
            "n_simulations": self.n_simulations,
            "return_ci": [round4(self.return_ci_lower), round4(self.return_ci_upper)],
            "sharpe_ci": [round4(self.sharpe_ci_lower), round4(self.sharpe_ci_upper)],
            "drawdown_ci": [round4(self.drawdown_ci_lower), round4(self.drawdown_ci_upper)],
            "p_value_return": round4(self.p_value_return),
            "observed_return": round4(self.observed_return),
            "observed_sharpe": round4(self.observed_sharpe),
        })
    }
}

const HOUR_MS: i64 = 3_600_000;
const DAY_MS: i64 = 24 * HOUR_MS;

/// The walk-forward engine. Pure window math — the backtest runs themselves
/// are delegated to a caller-supplied function so the Rust port stays
/// self-contained.
pub struct WalkForwardEngine;

impl WalkForwardEngine {
    /// Compute the rolling train/validate windows over the candle span.
    /// Returns per-window bounds; a caller runs each window's backtest.
    pub fn windows(
        candles: &[i64],
        train_days: i64,
        validate_days: i64,
        step_days: i64,
    ) -> Vec<(i64, i64, i64, i64)> {
        let (Some(first), Some(last)) = (candles.first(), candles.last()) else {
            return vec![];
        };
        let train_ms = train_days * DAY_MS;
        let validate_ms = validate_days * DAY_MS;
        let step_ms = step_days * DAY_MS;
        let last_ts = *last;
        let mut out = Vec::new();
        let mut window_start = *first;
        loop {
            let train_start = window_start;
            let train_end = train_start + train_ms;
            let validate_start = train_end;
            let validate_end = validate_start + validate_ms;
            if validate_end > last_ts {
                break;
            }
            out.push((train_start, train_end, validate_start, validate_end));
            window_start += step_ms;
        }
        out
    }
}

/// Monte Carlo bootstrap simulation.
pub fn run_monte_carlo(
    equity_curve: &[(i64, f64)],
    n_simulations: usize,
    confidence: f64,
    seed: u64,
) -> MonteCarloResult {
    if equity_curve.len() < 2 || n_simulations == 0 {
        // C1: zero simulations must return an empty result, not index an empty
        // vector.
        return MonteCarloResult {
            n_simulations: 0,
            return_ci_lower: 0.0,
            return_ci_upper: 0.0,
            sharpe_ci_lower: 0.0,
            sharpe_ci_upper: 0.0,
            drawdown_ci_lower: 0.0,
            drawdown_ci_upper: 0.0,
            p_value_return: 1.0,
            observed_return: 0.0,
            observed_sharpe: 0.0,
        };
    }

    let mut rng = rand::rng_from_seed(seed);
    let observed_returns = compute_returns(equity_curve);
    let observed_total_return = if equity_curve[0].1 > 0.0 {
        equity_curve[equity_curve.len() - 1].1 / equity_curve[0].1 - 1.0
    } else {
        0.0
    };
    let observed_sharpe = compute_sharpe(&observed_returns);

    let n = observed_returns.len();
    let mut sim_returns = Vec::with_capacity(n_simulations);
    let mut sim_sharpes = Vec::with_capacity(n_simulations);
    let mut sim_drawdowns = Vec::with_capacity(n_simulations);

    for _ in 0..n_simulations {
        let resampled: Vec<f64> = (0..n).map(|_| observed_returns[rng.next() % n]).collect();
        let mut sim_equity = equity_curve[0].1;
        let mut sim_curve = vec![(equity_curve[0].0, sim_equity)];
        for (i, ret) in resampled.iter().enumerate() {
            sim_equity *= 1.0 + ret;
            let ts = equity_curve[(i + 1).min(equity_curve.len() - 1)].0;
            sim_curve.push((ts, sim_equity));
        }
        let initial = sim_curve[0].1;
        let sim_total_return = if initial > 0.0 {
            sim_curve[sim_curve.len() - 1].1 / initial - 1.0
        } else {
            0.0
        };
        sim_returns.push(sim_total_return);
        sim_sharpes.push(compute_sharpe(&resampled));
        sim_drawdowns.push(compute_max_drawdown(&sim_curve));
    }

    let alpha = (1.0 - confidence) / 2.0;
    sim_returns.sort_by(|a, b| a.partial_cmp(b).unwrap());
    sim_sharpes.sort_by(|a, b| a.partial_cmp(b).unwrap());
    sim_drawdowns.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let lower_idx = ((alpha * n_simulations as f64) as usize).max(0);
    let upper_idx = (((1.0 - alpha) * n_simulations as f64) as usize).min(n_simulations.saturating_sub(1));
    let p_value = sim_returns.iter().filter(|r| **r >= observed_total_return).count() as f64 / n_simulations as f64;

    MonteCarloResult {
        n_simulations,
        return_ci_lower: sim_returns[lower_idx],
        return_ci_upper: sim_returns[upper_idx],
        sharpe_ci_lower: sim_sharpes[lower_idx],
        sharpe_ci_upper: sim_sharpes[upper_idx],
        drawdown_ci_lower: sim_drawdowns[lower_idx],
        drawdown_ci_upper: sim_drawdowns[upper_idx],
        p_value_return: p_value,
        observed_return: observed_total_return,
        observed_sharpe,
    }
}

/// Bonferroni multiple testing correction.
pub fn bonferroni_correction(p_value: f64, n_tests: usize) -> f64 {
    (p_value * n_tests as f64).min(1.0)
}

/// Simple returns from an equity curve.
pub fn compute_returns(equity_curve: &[(i64, f64)]) -> Vec<f64> {
    let mut returns = Vec::new();
    for i in 1..equity_curve.len() {
        let prev = equity_curve[i - 1].1;
        let curr = equity_curve[i].1;
        if prev > 0.0 {
            returns.push((curr - prev) / prev);
        }
    }
    returns
}

/// Annualized Sharpe from a return series (assumes hourly → ~8760/yr).
pub fn compute_sharpe(returns: &[f64]) -> f64 {
    if returns.is_empty() {
        return 0.0;
    }
    let mean = returns.iter().sum::<f64>() / returns.len() as f64;
    let variance = returns.iter().map(|r| (r - mean).powi(2)).sum::<f64>() / returns.len() as f64;
    let std = if variance > 0.0 { variance.sqrt() } else { 0.0 };
    if std == 0.0 {
        return 0.0;
    }
    (mean / std) * 8760.0f64.sqrt()
}

/// Max drawdown from an equity curve.
pub fn compute_max_drawdown(equity_curve: &[(i64, f64)]) -> f64 {
    let mut peak = 0.0;
    let mut max_dd = 0.0;
    for (_, eq) in equity_curve {
        if *eq > peak {
            peak = *eq;
        }
        if peak > 0.0 {
            let dd = (peak - eq) / peak;
            if dd > max_dd {
                max_dd = dd;
            }
        }
    }
    max_dd
}

fn round4(v: f64) -> f64 {
    (v * 10_000.0).round() / 10_000.0
}

/// A tiny deterministic PRNG (xorshift64) so Monte Carlo is reproducible
/// without pulling in a rand dependency.
mod rand {
    pub struct Xorshift64(u64);
    impl Xorshift64 {
        pub fn next(&mut self) -> usize {
            self.0 ^= self.0 << 13;
            self.0 ^= self.0 >> 7;
            self.0 ^= self.0 << 17;
            self.0 as usize
        }
    }
    pub fn rng_from_seed(seed: u64) -> Xorshift64 {
        Xorshift64(if seed == 0 { 0x9E3779B97F4A7C15 } else { seed })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn walk_forward_windows_slide() {
        // 100 hourly candles starting at 0 → span ~4.17 days.
        let candles: Vec<i64> = (0..100).map(|i| i * HOUR_MS).collect();
        let windows = WalkForwardEngine::windows(&candles, 1, 1, 1);
        assert!(!windows.is_empty());
        // Each window: train [start, start+1d), validate [start+1d, start+2d).
        let (ts, te, vs, ve) = windows[0];
        assert_eq!(te - ts, DAY_MS);
        assert_eq!(vs, te);
        assert_eq!(ve - vs, DAY_MS);
        // Windows step by 1 day.
        assert_eq!(windows[1].0 - ts, DAY_MS);
    }

    #[test]
    fn walk_forward_windows_respects_span() {
        // Too short: 1 day of candles can't fit train+validate of 1d each.
        let candles: Vec<i64> = (0..23).map(|i| i * HOUR_MS).collect();
        let windows = WalkForwardEngine::windows(&candles, 1, 1, 1);
        assert!(windows.is_empty());
    }

    #[test]
    fn bonferroni_correction_caps_at_one() {
        assert_eq!(bonferroni_correction(0.01, 5), 0.05);
        assert_eq!(bonferroni_correction(0.3, 10), 1.0);
    }

    #[test]
    fn sharpe_and_max_drawdown_basic() {
        let curve = vec![(0, 100.0), (1, 110.0), (2, 95.0), (3, 105.0)];
        let returns = compute_returns(&curve);
        assert_eq!(returns.len(), 3);
        assert!((returns[0] - 0.1).abs() < 1e-9);
        let sharpe = compute_sharpe(&returns);
        assert!(sharpe > 0.0);
        // Max drawdown: peak 110 → trough 95 → (110-95)/110 ≈ 0.136.
        let dd = compute_max_drawdown(&curve);
        assert!((dd - (110.0 - 95.0) / 110.0).abs() < 1e-9);
    }

    #[test]
    fn monte_carlo_runs_deterministically() {
        let curve: Vec<(i64, f64)> = (0..50).map(|i| (i, 100.0 * (1.0 + i as f64 * 0.002))).collect();
        let r1 = run_monte_carlo(&curve, 200, 0.95, 42);
        let r2 = run_monte_carlo(&curve, 200, 0.95, 42);
        assert_eq!(r1.return_ci_lower, r2.return_ci_lower);
        assert_eq!(r1.return_ci_upper, r2.return_ci_upper);
        assert_eq!(r1.p_value_return, r2.p_value_return);
        assert_eq!(r1.n_simulations, 200);
        assert!(r1.observed_return > 0.0);
    }

    #[test]
    fn monte_carlo_zero_simulations_does_not_panic() {
        // C1 regression: 0 simulations must return an empty result, not index
        // into an empty vector.
        let curve: Vec<(i64, f64)> = vec![(0, 100.0), (1, 101.0)];
        let r = run_monte_carlo(&curve, 0, 0.95, 1);
        assert_eq!(r.n_simulations, 0);
        assert_eq!(r.return_ci_lower, 0.0);
        assert_eq!(r.p_value_return, 1.0);
    }
}
